provider/azure/environ_network.go

```diff
@@ -133,11 +133,163 @@ func (*azureEnviron) SSHAddresses(
+	}
+	subnetIDToCIDR := make(map[string]string)
+	for _, sub := range allSubnets {
+		subnetIDToCIDR[sub.ProviderId.String()] = sub.CIDR
+	}
+
+	nicClient := azurenetwork.InterfacesClient{BaseClient: env.network}
+	instIfaceMap, err := instanceNetworkInterfaces(ctx, env.resourceGroup, nicClient)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var (
+		res        = make([]network.InterfaceInfos, len(instanceIDs))
+		matchCount int
+	)
+
+	for resIdx, instID := range instanceIDs {
+		azInterfaceList, found := instIfaceMap[instID]
+		if !found {
+			continue
+		}
+
+		matchCount++
+		res[resIdx] = mapAzureInterfaceList(azInterfaceList, subnetIDToCIDR)
+	}
+
+	if matchCount == 0 {
+		return nil, environs.ErrNoInstances
+	} else if matchCount < len(instanceIDs) {
+		return res, environs.ErrPartialInstances
+	}
+
+	return res, nil
+}
+
+func mapAzureInterfaceList(in []azurenetwork.Interface, subnetIDToCIDR map[string]string) network.InterfaceInfos {
+	var out = make(network.InterfaceInfos, len(in))
+
+	for idx, azif := range in {
+		ni := network.InterfaceInfo{
+			DeviceIndex:   idx,
+			Disabled:      false,
+			NoAutoStart:   false,
+			InterfaceType: network.EthernetDevice,
+			Origin:        network.OriginProvider,
+		}
+
+		if azif.MacAddress != nil {
+			ni.MACAddress = network.NormalizeMACAddress(*azif.MacAddress)
+		}
+		if azif.ID != nil {
+			ni.ProviderId = network.Id(*azif.ID)
+		}
+
+		if azif.InterfacePropertiesFormat == nil || azif.IPConfigurations == nil {
+			out[idx] = ni
+			continue
+		}
+
+		for _, ipConf := range *azif.IPConfigurations {
+			if ipConf.InterfaceIPConfigurationPropertiesFormat == nil {
+				continue
+			}
+
+			isPrimary := ipConf.Primary != nil && *ipConf.Primary
+
+			// NOTE(achilleasa): azure does not seem to include
+			// shadow IPs (or any public IP for that matter) when
+			// querying network interfaces. If we do encounter a
+			// public IP make sure we include it both as an address
+			// and as a shadow IP.
+			if ipConf.PublicIPAddress != nil && ipConf.PublicIPAddress.PublicIPAddressPropertiesFormat != nil && ipConf.PublicIPAddress.IPAddress != nil {
+				var cfgMethod = network.ConfigDHCP
+				if ipConf.PublicIPAddress.PublicIPAllocationMethod == azurenetwork.Static {
+					cfgMethod = network.ConfigStatic
+				}
+
+				providerAddr := network.NewMachineAddress(
+					*ipConf.PublicIPAddress.IPAddress,
+					network.WithScope(network.ScopePublic),
+					network.WithConfigType(cfgMethod),
+				).AsProviderAddress()
+
+				// If this a primary address make sure it appears
+				// at the top of the address lists.
+				if isPrimary {
+					ni.ShadowAddresses = append(network.ProviderAddresses{providerAddr}, ni.ShadowAddresses...)
+					ni.Addresses = append(network.ProviderAddresses{providerAddr}, ni.Addresses...)
+					ni.ConfigType = cfgMethod
+				} else {
+					ni.ShadowAddresses = append(ni.ShadowAddresses, providerAddr)
+					ni.Addresses = append(ni.Addresses, providerAddr)
```

Comment:
We don't do this for AWS, I believe we will get an logged message in the instance-poller for this address anyway, because there is no attendant subnet.

I think I would prefer just to set it as a shadow address, so we can handle it specifically when that part is implemented.

Reply:
Sure thing. I will push a commit to clean this up.

For AWS in particular they give us a list of the private addresses plus (in the same payload) the shadow ones which we know that they are never visible to the machine itself. I wasn't sure if that is also the case for azure so I opted to add it to both collections (also, see my note above about azure not returning the public IPs in the response payloads).